core/include/cubos/core/reflection/type.hpp

```diff
@@ -0,0 +1,131 @@
+        static void destroy(Type& type);
+
+        /// @brief Returns the name of the type.
+        /// @return Name of the type.
+        const std::string& name() const;
+
+        /// @brief Adds the given trait to the type.
+        ///
+        /// Aborts if the trait is already present in the type, or if its not move or copy
+        /// constructible. Fails to compile if the trait's type is not reflectable.
+        ///
+        /// @tparam T %Trait type.
+        /// @param trait %Trait value.
+        /// @return Reference to this type, for chaining.
+        template <typename T>
+        Type& with(T trait)
+        {
+            return this->with(Type::id<T>(), new T(trait), [](void* trait) { delete static_cast<T*>(trait); });
+        }
+
+        /// @brief Returns whether the type has the given trait.
+        ///
+        /// Fails to compile if the trait's type is not reflectable.
+        ///
+        /// @tparam T %Trait type.
+        /// @return Whether the type has the given trait.
+        template <typename T>
+        bool has() const
+        {
+            return this->has(Type::id<T>());
+        }
+
+        /// @brief Returns the given trait of the type.
+        ///
+        /// Aborts if the type does not have the given trait.
+        /// Fails to compile if the trait's type is not reflectable.
+        ///
+        /// @tparam T %Trait type.
+        /// @return Reference to the trait.
+        template <typename T>
+        const T& get() const
+        {
+            return *static_cast<const T*>(this->get(Type::id<T>()));
+        }
+
+    private:
+        ~Type();
+
+        /// @brief Constructs with the given name.
+        /// @param name Name of the type.
+        explicit Type(std::string name);
+
+        /// @brief Adds the given trait to the type.
+        /// @param id Identifies the type of the trait.
+        /// @param trait %Trait value to be moved into the type.
+        /// @param deleter Used to delete the trait when the type is destroyed.
+        /// @return Reference to this type, for chaining.
+        Type& with(uintptr_t id, void* trait, void (*deleter)(void*));
+
+        /// @brief Returns whether the type has the given trait.
+        /// @param id Identifies the type of the trait.
+        /// @return Whether the type has the given trait.
+        bool has(uintptr_t id) const;
+
+        /// @brief Returns the given trait of the type.
+        ///
+        /// Aborts if the type does not have the given trait.
+        ///
+        /// @param id Identifies the type of the trait.
+        /// @return Pointer to the trait.
+        const void* get(uintptr_t id) const;
+
+        /// @brief Gets an unique identifier for the given type.
+        /// @note This function is used as an alternative to `std::type_index`, which would require
+        /// including the `<typeindex>` header.
+        /// @tparam T %Trait type.
+        /// @return Unique identifier for the given type.
```

Comment:
Is the id of the trait or of the type? If I understood correctly its the id of the type of the trait, but as this is inside the class type we should be careful using language like `given type`. This complaint extends to most of this file

Reply:
Its the id of  the type of the trait - cleared it up.